cpp/arcticdb/async/async_store.hpp

```diff
@@ -75,278 +76,266 @@ class AsyncStore : public Store {
-        res.reserve(keys.size());
-        for (auto itr = keys.cbegin(); itr < keys.cend(); ++itr) {
-            res.push_back(async::submit_io_task(KeyExistsTask(itr, library_)));
-        }
-        return res;
-    }
 
-    folly::Future<std::vector<VariantKey>> batch_write(
-        std::vector<std::pair<PartialKey, SegmentInMemory>> &&key_seg_pairs,
-        const std::shared_ptr<DeDupMap> &de_dup_map,
-        const BatchWriteArgs &args) override {
-        auto key_segments = std::move(key_seg_pairs);
-        std::vector<folly::Future<VariantKey>> futs;
-        futs.reserve(key_seg_pairs.size());
-        std::size_t write_count = args.lib_write_count == 0 ? 16ULL : args.lib_write_count;
-
-        auto encode_futs = folly::window(key_segments, [*this](auto &&ks) {
-            auto [key, seg] = std::forward<decltype(ks)>(ks);
-            return async::submit_cpu_task(
-                EncodeAtomTask(std::move(key),
-                               ClockType::nanos_since_epoch(),
-                               std::move(seg),
-                               codec_,
-                               encoding_version_));
-        }, write_count);
-
-        for (folly::Future<storage::KeySegmentPair>& encode_fut : encode_futs) {
-            futs.emplace_back(
-                std::move(encode_fut).thenValue([de_dup_map](auto &&ks) -> std::pair<VariantKey, std::optional<Segment>> {
-                        auto key_seg = std::forward<decltype(ks)>(ks);
-                        return lookup_match_in_dedup_map(de_dup_map, std::move(key_seg));
-                    })
-                    .via(&async::io_executor()).thenValue([lib = library_](auto &&item) {
-                        auto key_opt_segment = std::forward<decltype(item)>(item);
-                        if (key_opt_segment.second)
-                            lib->write(Composite<storage::KeySegmentPair>({VariantKey{key_opt_segment.first},
-                                                                           std::move(*key_opt_segment.second)}));
-
-                        return key_opt_segment.first;
-                    }));
-        }
+    folly::Future<SliceAndKey> async_write(
+            folly::Future<std::tuple<PartialKey, SegmentInMemory, pipelines::FrameSlice>> &&input_fut,
+            const std::shared_ptr<DeDupMap> &de_dup_map) override {
+        using KeyOptSegment = std::pair<VariantKey, std::optional<Segment>>;
+        return std::move(input_fut).thenValue([this] (auto&& input) {
+            auto [key, seg, slice] = std::move(input);
+            auto key_seg = EncodeAtomTask{
+                std::move(key),
+                ClockType::nanos_since_epoch(),
+                std::move(seg),
+                codec_,
+                encoding_version_}();
+            return std::make_tuple<VariantKey, Segment, pipelines::FrameSlice>(std::move(key_seg.variant_key()), std::move(key_seg.segment()), std::move(slice));
```

Comment:
This could just be `std::pair<KeySegmentPair, FrameSlice>`, as the key segment pair is just reconstructed in the next lambda anyway.